test_fms/random_numbers/test_random_numbers.F90

```diff
@@ -0,0 +1,279 @@
+  call test_samples_iter(stream, test_sample_2d, n0_2d)
+end subroutine test_getRandomNumbers_dispatch
+
+!> Run the requested test using progressively larger sample sizes, until ten
+!> passing samples have been drawn
+subroutine test_samples_iter(stream, test, n0)
+  abstract interface
+    !> Abstract interface for test_sample_1d and test_sample_2d
+    function test_sample(stream, n)
+      import :: randomNumberStream
+      type(randomNumberStream), intent(inout) :: stream
+      integer, intent(in) :: n
+      logical :: test_sample
+    end function
+  end interface
+
+  integer, parameter :: required_passes = 10 !> Number of samples that must pass the test
+
+  type(randomNumberStream), intent(inout) :: stream !> Random number stream
+  procedure(test_sample) :: test !> Function which draws a random sample and tests it
+  integer, intent(in) :: n0 !> Initial sample size
+
+  real(k) :: x !> Sample for 0D test
+  integer :: n !> Sample size for 1D or 2D test
+  integer :: pass_counter !> Number of test passes
+
+  ! 0D case
+  ! Draw a scalar and check that it's within [0,1]
+
+  call getRandomNumbers(stream, x)
+  call check_bounds(x)
+
+  ! 1D and 2D cases
+  ! Attempt to draw ten samples for which the first 1,000 moments are within one
+  ! standard deviation of their expected values for the uniform distribution on
+  ! [0,1]. Draw progressively larger samples until this condition has been fulfilled.
+
+  n = n0
+  pass_counter = 0
+
+  do while (pass_counter .lt. required_passes)
+    if (test(stream, n)) then
+      pass_counter = pass_counter + 1
+    endif
+
+    n = n * 11 / 10
+  enddo
+end subroutine test_samples_iter
+
+!> Draw a random sample and test its values and moments (1D)
+function test_sample_1d(stream, n)
+  type(randomNumberStream), intent(inout) :: stream !> Random number stream
+  integer, intent(in) :: n !> Length of the 1D sample vector
+  logical :: test_sample_1d !> True if the test passes for this sample
+  real(k) :: v(n) !> Sample vector
+  integer :: i !> Indices into v(:)
+
+  call getRandomNumbers(stream, v)
+
+  do i = 1, n
+    call check_bounds(v(i))
+  enddo
+
+  test_sample_1d = compare_sample_moments(v)
+end function test_sample_1d
+
+!> Draw a random sample and test its values and moments (2D)
+function test_sample_2d(stream, n)
+  type(randomNumberStream), intent(inout) :: stream !> Random number stream
+  integer, intent(in) :: n !> Dimensions of the 2D sample array
+  logical :: test_sample_2d !> True if the test passes for this sample
+  real(k) :: arr(n,n) !> Sample array
+  integer :: i, j !> Indices into arr(:,:)
+
+  call getRandomNumbers(stream, arr)
+
+  do j = 1, n
+    do i = 1, n
+      call check_bounds(arr(i, j))
+    enddo
+  enddo
+
+  test_sample_2d = compare_sample_moments(reshape(arr, [size(arr)]))
+end function test_sample_2d
+
+!> Check that the first 1,000 moments of a sample are within one standard
+!> deviation of their expected values
```

Comment:
why one standard deviation?  do we know what the distribution function looks like?

Reply:
One standard deviation was an empirical choice, based on the observation that the test passes within a couple seconds. But if we need the test to run faster, we could use a more lenient condition like 1.5 or 2 standard deviations.

I believe the central limit theorem implies that the distribution of sample moments should be approximately Gaussian, at least in the limit as the sample size approaches infinity.